lib/src/shared/main/java/com/couchbase/lite/internal/AbstractExecutionService.java

```diff
@@ -75,33 +105,113 @@ public String toString() {
-        private final Executor executor;
+        @NonNull
+        private final ThreadPoolExecutor executor;
+
+        @GuardedBy("this")
+        @NonNull
+        private final Queue<InstrumentedTask> pendingTasks = new LinkedList<>();
+
+        // a non-null stop latch is the flag that this executor has been stopped
+        @GuardedBy("this")
+        @Nullable
         private CountDownLatch stopLatch;
+
+        @GuardedBy("this")
         private int running;
 
-        private ConcurrentExecutor(Executor executor) { this.executor = executor; }
+        @GuardedBy("this")
+        private boolean needsRestart;
+
+        ConcurrentExecutor(@NonNull ThreadPoolExecutor executor) {
+            Preconditions.checkArgNotNull(executor, "executor");
+            this.executor = executor;
+        }
 
+        /**
+         * Schedule a task for concurrent execution.
+         * There are absolutely no guarantees about execution order, on this executor,
+         * particularly once it fails back to using the pending task queue.
+         * If there is insufficient room to schedule the task, safely, on the underlying
+         * executor, the task is added to the pendingTask queue and executed when space
+         * is available.
+         * This method may throw a <code>RejectedExecutionException</code> if the underlying
+         * executor's resources are completely exhausted even though <code>spaceAvailable</code>
+         * returns true.
+         *
+         * @param task a task for concurrent execution.
+         * @throws ExecutorClosedException    if the executor has been stopped
+         * @throws RejectedExecutionException if the underlying executor rejects the task
+         */
         @Override
-        public synchronized void execute(@NonNull Runnable task) {
-            if (stopLatch != null) { throw new ExecutorClosedExeception("Executor has been stopped"); }
+        public void execute(@NonNull Runnable task) {
+            Preconditions.checkArgNotNull(task, "task");
 
-            try {
-                executor.execute(new InstrumentedTask(task, this::finishTask));
-                running++;
-            }
-            catch (RejectedExecutionException e) {
-                dumpServiceState(executor, e, "size: " + running);
-                throw e;
+            final int pendingTaskCount;
+            synchronized (this) {
+                if (stopLatch != null) { throw new ExecutorClosedException("Executor has been stopped"); }
+
+                if (spaceAvailable()) {
+                    if (needsRestart) { restartQueue(); }
+
+                    executeTask(new InstrumentedTask(task, this::finishTask));
+
+                    return;
+                }
+
+                pendingTasks.add(new InstrumentedTask(task));
+
+                pendingTaskCount = pendingTasks.size();
+                if (needsRestart || (pendingTaskCount == 1)) { restartQueue(); }
```

Comment:
Is it a good idea to restart the queue when no space is available?  Or will there possibly be space available now after a few lines from when you checked before?

Reply:
Yeah... 
1) the contract is that the serial queue is not controlled by `spaceAvailable`.  The concurrent executor can fill all but the last 62 slots: 64 - 2 (for the serial queue).  That needs documentation
2) `restartQueue` will attempt to restart the queue, but will fail, if the executor is truly swamped.  It that case, it will continue to be stalled...